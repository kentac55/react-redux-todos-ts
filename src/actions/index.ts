import actionCreatorFactory from 'typescript-fsa'
import { AsyncOpResult, Todo, VisibilityFilter } from '../types'

const actionCreator = actionCreatorFactory()

type AddTodoProp = {
  text: string
}
export const addTodo = actionCreator.async<
  AddTodoProp,
  AsyncOpResult<Todo>,
  Error
>('AddTodo')

type ToggleTodoProp = {
  todo: Todo
}
export const toggleTodo = actionCreator.async<
  ToggleTodoProp,
  AsyncOpResult<Todo>,
  Error
>('ToggleTodo')

export const setVisibilityFilter = actionCreator<VisibilityFilter>(
  'SetVisibilityFilter'
)

type GetTodoProp = {
  id: number
}
export const getTodo = actionCreator.async<
  GetTodoProp,
  AsyncOpResult<Todo>,
  Error
>('GetTodo')

type GetTodosProp = void
export const getTodos = actionCreator.async<
  GetTodosProp,
  AsyncOpResult<Todo[]>,
  Error
>('GetTogos')

type InitProp = void
export const initApp = actionCreator.async<
  InitProp,
  AsyncOpResult<Todo[]>,
  Error
>('InitApp')
