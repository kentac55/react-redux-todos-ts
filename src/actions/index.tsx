import actionCreatorFactory from 'typescript-fsa'
import { VisibilityFilterKinds } from '../types'

const actionCreator = actionCreatorFactory()
export const addTodoAction = actionCreator<string>('ADD')
export const toggleTodoAction = actionCreator<number>('TOGGLE')

export const setVisibilityFilterAction = actionCreator<VisibilityFilterKinds>(
  'SET'
)
