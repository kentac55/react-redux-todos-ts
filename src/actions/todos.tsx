let nextTodoId = 0
export const addTodoAction = (text: string) =>
  ({
    type: 'ADD',
    id: nextTodoId++,
    text,
  } as const)

export const toggleTodoAction = (id: number) =>
  ({
    type: 'TOGGLE',
    id,
  } as const)
