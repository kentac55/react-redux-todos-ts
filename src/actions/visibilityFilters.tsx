import { VisibilityFilterKinds } from '../types'

export const setVisibilityFilterAction = (filter: VisibilityFilterKinds) =>
  ({
    type: 'SET_VISIBILITY_FILTER',
    filter,
  } as const)
