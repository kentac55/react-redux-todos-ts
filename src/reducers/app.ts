import { reducerWithInitialState } from 'typescript-fsa-reducers'
import { addTodo, initApp, getTodo, getTodos, toggleTodo } from '../actions'
import { AsyncOpResult } from '../types'

export type AppState = {
  loading: boolean
  loaded: boolean
  error: {
    type: string
    message: string | null
  } | null
}

const failedHandler = (_: unknown, { error }: { error: Error }): AppState => {
  return {
    loaded: true,
    loading: false,
    error: { type: error.name, message: error.message },
  }
}
const doneHandler = (
  state: AppState,
  { result }: { result: AsyncOpResult<unknown> }
): AppState => {
  if (result.type === 'Ok') {
    return {
      ...state,
      loaded: true,
      loading: false,
    }
  } else {
    return {
      loaded: true,
      loading: false,
      error: { type: result.type, message: result.message },
    }
  }
}
const startedHandler = (state: AppState): AppState => {
  return { ...state, loading: true }
}

export const appReducer = reducerWithInitialState({
  loaded: false,
  loading: false,
  error: null,
} as AppState)
  .cases(
    [addTodo.started, getTodo.started, initApp.started, toggleTodo.started],
    startedHandler
  )
  .case(getTodos.started, startedHandler)
  .cases(
    [addTodo.failed, getTodo.failed, initApp.failed, toggleTodo.failed],
    failedHandler
  )
  .case(getTodos.failed, failedHandler)
  .cases(
    [addTodo.done, getTodo.done, initApp.done, toggleTodo.done],
    doneHandler
  )
  .case(getTodos.done, doneHandler)
