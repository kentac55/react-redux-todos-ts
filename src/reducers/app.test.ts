import { Action } from 'redux'
import { appReducer, AppState } from './app'
import { addTodo, initApp, getTodo, getTodos, toggleTodo } from '../actions'

const id = 0
const text = 'hoge'
const todo = { id, text, completed: false }

const init = {
  loaded: false,
  loading: false,
  error: null,
}

const loadingState = {
  ...init,
  loading: true,
}

const error = Error('error')

const errorState = {
  ...init,
  loaded: true,
  error: { type: error.name, message: error.message },
}

const loadedState = {
  ...init,
  loaded: true,
  loading: false,
}

const serverError = { type: 'Internal', message: null } as const

it('should return initial state', () => {
  expect(appReducer(undefined, { type: undefined })).toEqual(init)
  expect(appReducer(undefined, { type: undefined })).toMatchSnapshot()
})

it.each<[Action, AppState]>([
  [addTodo.started({ text }), loadingState],
  [getTodo.started({ id }), loadingState],
  [initApp.started(), loadingState],
  [toggleTodo.started({ todo }), loadingState],
  [getTodos.started(), loadingState],
  [addTodo.failed({ params: { text }, error }), errorState],
  [getTodo.failed({ params: { id }, error }), errorState],
  [initApp.failed({ params: undefined, error }), errorState],
  [
    toggleTodo.failed({
      params: { todo },
      error,
    }),
    errorState,
  ],
  [getTodos.failed({ params: undefined, error }), errorState],
  [
    addTodo.done({
      params: { text },
      result: { type: 'Ok', contents: todo },
    }),
    loadedState,
  ],
  [
    getTodo.done({ params: { id }, result: { type: 'Ok', contents: todo } }),
    loadedState,
  ],
  [
    initApp.done({
      params: undefined,
      result: { type: 'Ok', contents: [todo] },
    }),
    loadedState,
  ],
  [
    toggleTodo.done({
      params: { todo },
      result: { type: 'Ok', contents: todo },
    }),
    loadedState,
  ],
  [
    getTodos.done({
      params: undefined,
      result: { type: 'Ok', contents: [todo] },
    }),
    loadedState,
  ],
  [
    addTodo.done({ params: { text }, result: serverError }),
    { loaded: true, loading: false, error: serverError },
  ],
])('should handle %s', (action, expectedState) => {
  expect(appReducer(init, action)).toEqual(expectedState)
  expect(appReducer(init, action)).toMatchSnapshot()
})
