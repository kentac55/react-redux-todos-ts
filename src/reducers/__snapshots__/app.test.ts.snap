// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle { type: 'AddTodo_DONE', payload: [Object] } 1`] = `
Object {
  "error": null,
  "loaded": true,
  "loading": false,
}
`;

exports[`should handle { type: 'AddTodo_DONE', payload: [Object] } 2`] = `
Object {
  "error": Object {
    "message": null,
    "type": "Internal",
  },
  "loaded": true,
  "loading": false,
}
`;

exports[`should handle { type: 'AddTodo_FAILED', payload: [Object], error: true } 1`] = `
Object {
  "error": Object {
    "message": "error",
    "type": "Error",
  },
  "loaded": true,
  "loading": false,
}
`;

exports[`should handle { type: 'AddTodo_STARTED', payload: [Object] } 1`] = `
Object {
  "error": null,
  "loaded": false,
  "loading": true,
}
`;

exports[`should handle { type: 'GetTodo_DONE', payload: [Object] } 1`] = `
Object {
  "error": null,
  "loaded": true,
  "loading": false,
}
`;

exports[`should handle { type: 'GetTodo_FAILED', payload: [Object], error: true } 1`] = `
Object {
  "error": Object {
    "message": "error",
    "type": "Error",
  },
  "loaded": true,
  "loading": false,
}
`;

exports[`should handle { type: 'GetTodo_STARTED', payload: [Object] } 1`] = `
Object {
  "error": null,
  "loaded": false,
  "loading": true,
}
`;

exports[`should handle { type: 'GetTodos_DONE', payload: [Object] } 1`] = `
Object {
  "error": null,
  "loaded": true,
  "loading": false,
}
`;

exports[`should handle { type: 'GetTodos_FAILED', payload: [Object], error: true } 1`] = `
Object {
  "error": Object {
    "message": "error",
    "type": "Error",
  },
  "loaded": true,
  "loading": false,
}
`;

exports[`should handle { type: 'GetTodos_STARTED', payload: undefined } 1`] = `
Object {
  "error": null,
  "loaded": false,
  "loading": true,
}
`;

exports[`should handle { type: 'InitApp_DONE', payload: [Object] } 1`] = `
Object {
  "error": null,
  "loaded": true,
  "loading": false,
}
`;

exports[`should handle { type: 'InitApp_FAILED', payload: [Object], error: true } 1`] = `
Object {
  "error": Object {
    "message": "error",
    "type": "Error",
  },
  "loaded": true,
  "loading": false,
}
`;

exports[`should handle { type: 'InitApp_STARTED', payload: undefined } 1`] = `
Object {
  "error": null,
  "loaded": false,
  "loading": true,
}
`;

exports[`should handle { type: 'ToggleTodo_DONE', payload: [Object] } 1`] = `
Object {
  "error": null,
  "loaded": true,
  "loading": false,
}
`;

exports[`should handle { type: 'ToggleTodo_FAILED', payload: [Object], error: true } 1`] = `
Object {
  "error": Object {
    "message": "error",
    "type": "Error",
  },
  "loaded": true,
  "loading": false,
}
`;

exports[`should handle { type: 'ToggleTodo_STARTED', payload: [Object] } 1`] = `
Object {
  "error": null,
  "loaded": false,
  "loading": true,
}
`;

exports[`should return initial state 1`] = `
Object {
  "error": null,
  "loaded": false,
  "loading": false,
}
`;
